Базы данных

lepfel opened the pull request:
Реализован доп. класс  Users для хранения полученных из базы пользователей, а также выполнены задания по получению данных из базы в MainJava.
Также выполнено задание по написанию метода сравнения баз данных без использования циклов в классе CompDatabase. Как работает метод можно увидеть в CompAppl.

lepfel committed:
Реализован доп. класс  Users для хранения полученных из базы пользователей, а также выполнены задания по получению данных из базы в MainJava.
Также выполнено задание по написанию метода сравнения баз данных без использования циклов в классе CompDatabase. Как работает метод можно увидеть в CompAppl.

iaulitin commented on src/ru/milandr/courses/pavlyuk/CompAppl.java:
```diff
@@ -0,0 +1,12 @@
+
+public class CompAppl {
+    public static void main(String[] args) {
+        boolean a = CompDatabase.compareData("jdbc:postgresql://localhost:5432/users_database",
```
Давай переименуем переменную.

iaulitin commented on src/ru/milandr/courses/pavlyuk/CompDatabase.java:
```diff
@@ -0,0 +1,76 @@
+package ru.milandr.courses.pavlyuk;
+
+
+import com.sun.prism.impl.Disposer;
```
Неиспользуемые импорты не почистила.

iaulitin commented on src/ru/milandr/courses/pavlyuk/CompDatabase.java:
```diff
@@ -0,0 +1,76 @@
+             Connection con2 = DriverManager.getConnection(db2Link,
+                     "postgres", "postgres");
+             Statement smt = con.createStatement(ResultSet.TYPE_SCROLL_SENSITIVE,
+                     ResultSet.CONCUR_UPDATABLE);
```
А зачем UPDATABLE, если ты не обновляешь таблицы?

iaulitin commented on src/ru/milandr/courses/pavlyuk/CompDatabase.java:
```diff
@@ -0,0 +1,76 @@
+public class CompDatabase {
+
+
+    public static boolean compareData(String db1Link, String db2Link, String table1Name,
```
Разнеси, пожалуйста, аргументы по строкам.

iaulitin replied:
Нет проверки входных аргументов на корректность.

iaulitin commented on src/ru/milandr/courses/pavlyuk/MainJava.java:
```diff
@@ -0,0 +1,88 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class MainJava {
```
Давай переименуем в HomeTask4.

iaulitin commented on src/ru/milandr/courses/pavlyuk/CompAppl.java:
```diff
@@ -0,0 +1,12 @@
+package ru.milandr.courses.pavlyuk;
+
+public class CompAppl {
```
Давай переименуем в HomeTask5.

iaulitin commented on src/ru/milandr/courses/pavlyuk/MainJava.java:
```diff
@@ -0,0 +1,88 @@
+    public static void main(String[] args) {
+
+        List<User> Users = new ArrayList<>();
+        List<User> AnteRoadUsers = new ArrayList<>();
```
Что такое AnteRoad? :)

iaulitin replied:
Переменные не должны имя, начинающееся с большой буквы.

iaulitin commented on src/ru/milandr/courses/pavlyuk/MainJava.java:
```diff
@@ -0,0 +1,88 @@
+            //2. Вывести всех пользователей, проживающих по адресу P.O. Box 677, 8665 Ante Road.
+            // Для того, чтобы это сделать необходимо самостоятельно освоить такую SQL-конструкцию как JOIN.
+
+            rs = smt.executeQuery("SELECT *\n" +
```
Символы переноса строки можно не писать в SQL. Они же нужны только тебе для удобства, а ты и так уже разбиваешь запрос на строки, которые затем склеиваешь.

iaulitin commented on src/ru/milandr/courses/pavlyuk/MainJava.java:
```diff
@@ -0,0 +1,88 @@
+                rs = smt.executeQuery("SELECT address\n" +
+                        "FROM users RIGHT JOIN addresses\n" +
+                        "ON users.address_id = addresses.id\n" +
+                        "WHERE first_name is NULL");
```
В других запросах ты ставишь символ ';' в конце, а в этом не ставишь. Давай сделаем единообразно.

iaulitin commented on src/ru/milandr/courses/pavlyuk/User.java:
```diff
@@ -0,0 +1,48 @@
+package ru.milandr.courses.pavlyuk;
+
+public class User {
+    private int  id;
```
Отсутствует форматирование.

iaulitin commented on src/ru/milandr/courses/pavlyuk/User.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @Override
+    public String toString() {
```
toString() обычно пишут в самом конце класса.

iaulitin commented on src/ru/milandr/courses/pavlyuk/User.java:
```diff
@@ -0,0 +1,48 @@
+    private int  id;
+    private String firstName;
+    private String lastName;
+    private int addressID;
```
Переменная не соответствует конвенции camelCase. Должно быть "addressId".

iaulitin replied:
Осталось еще в одном месте, но не смертельно. Просто запомни, что этой конвенции именования стоит придерживаться, если хочешь, чтобы другие разработчики легче читали твой код.

iaulitin commented on src/ru/milandr/courses/pavlyuk/MainJava.java:
```diff
@@ -0,0 +1,88 @@
+import java.util.stream.Collectors;
+
+public class MainJava {
+    public static void main(String[] args) {
```
Весь код в main. Давай логику унесем в отдельный класс, в котором сделаем по методу на каждый пункт задания.
Заодно зарефакторим выборку данных.

lepfel committed:
Реализован доп. класс  Users для хранения полученных из базы пользователей, а также выполнены задания по получению данных из базы в MainJava.
Также выполнено задание по написанию метода сравнения баз данных без использования циклов в классе CompDatabase. Как работает метод можно увидеть в CompAppl.

Mobilnik commented:
Хорошо. Стало почти совсем хорошо.
Теперь давай перейдем  к следующему шагу.

Насчет ДЗ4
1. Пробегись еще раз по коду и проверь, что camelCase-именование соблюдено. Местами остались огрехи.
2. В 4х методах из 5 ты одинаково выбираешь пользователей в ResultSet, затем бежишь по ResultSet. Есть такой принцип - DRY - don't repeat yourself. Обычно среда разработки подсказывает, что у тебя код дублируется. Повторяющиеся куски функционала следует выносить в отдельные функции. Сделай, пожалуйста. 
3. При сортировке списка пользователей ты со скобочками переборщила.
4. Почитай про JavaDoc. Оформи комментарии к методам в виде JavaDoc. можно на русском.

Насчет ДЗ5
1. Замечания по форматированию (операции во вложенных лямбдах было бы неплохо вынести на новые строки)
2. Слушай, а зачем тебе выполнять select *? Если в таблице 100 колонок, а тебе нужно сравнить только по одной, то ты получишь большой оверхэд и по чтению из БД (с диска компьютера) и по затраченной памяти. 
3. Помнишь, мы говорили о том, что не нужно лишний код заворачивать в try-catch? Вот в твоем вложенном try-catch ты допускаешь ошибку. Почитай, в каких случаях ResultSet может выбросить SQLException и сверься с твоим сообщением, выводимым при поимке ошибки.

lepfel committed:
Учтены правки из review 2 для ДЗ4 и ДЗ5.
